features/steps/steps.py

```diff
@@ -283,4 +329,30 @@ def step_impl(context, representation_id):
                     errors.append(representation_shape_error(inst, representation_id))
     
     handle_errors(context, errors)
-    
+
+@then("The {entity} must have no duplicate points")
+def step_impl(context, entity):
+    if getattr(context, 'applicable', True):
+        errors = []
+        precision = get_sending_system_precision(context)
+        for instance in context.instances:
+            points = get_entity_point_components(instance)
+            points_coordinates = [point.Coordinates for point in points]
+            points_coordinates_rounded = []
+            for point_coordinates in points_coordinates:
+                point_coordinates_rounded = tuple(round(point / precision) * precision for point in point_coordinates)
+                points_coordinates_rounded.append(point_coordinates_rounded)
+            if len(points_coordinates_rounded) != len(set(points_coordinates_rounded)):
+                errors.append(polyobject_structure_error(instance, points, 'duplicate_points'))
+    handle_errors(context, errors)
+
+
+@then("The {entity} first and last point must be identical by reference")
+def step_impl(context, entity):
+    if getattr(context, 'applicable', True):
+        errors = []
+        for instance in context.instances:
+            points = get_entity_point_components(instance)
+            if str(points[0]) != str(points[-1]):
```

Comment:
`points[0] == points[-1]` is ok for this. For instances of file it checks whether they are the same instance.

Same for is_a() with/without subtypes we should probably create section in the wiki on entity comparison.

- `points[0] == points[-1]` same instance
- `points[0].get_info(recursive=True, include_identifier=False) == points[-1].get_info(recursive=True, include_identifier=False)` same set of attribute values (recursively).

Reply:
Done. :)